src/hash_map.c

```diff
@@ -0,0 +1,562 @@
+static rcutils_ret_t hash_map_check_and_grow_map(rcutils_hash_map_t * hash_map)
+{
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+  if (hash_map->impl->size >= (LOAD_FACTOR * hash_map->impl->capacity)) {
+    size_t new_capacity = 2 * hash_map->impl->capacity;
+    rcutils_array_list_t * new_map = NULL;
+
+    ret = hash_map_allocate_new_map(&new_map, new_capacity, &hash_map->impl->allocator);
+    if (RCUTILS_RET_OK != ret) {
+      return ret;
+    }
+
+    for (size_t map_index = 0;
+      map_index < hash_map->impl->capacity && RCUTILS_RET_OK == ret;
+      ++map_index)
+    {
+      rcutils_array_list_t * bucket = &(hash_map->impl->map[map_index]);
+      // Is this a valid bucket with entries
+      if (NULL != bucket->impl) {
+        size_t bucket_size = 0;
+        ret = rcutils_array_list_get_size(bucket, &bucket_size);
+        if (RCUTILS_RET_OK != ret) {
+          return ret;
+        }
+
+        for (size_t bucket_index = 0;
+          bucket_index < bucket_size && RCUTILS_RET_OK == ret;
+          ++bucket_index)
+        {
+          rcutils_hash_map_entry_t * entry = NULL;
+          ret = rcutils_array_list_get(bucket, bucket_index, &entry);
+          if (RCUTILS_RET_OK == ret) {
+            size_t new_index = entry->hashed_key % new_capacity;
+            ret = hash_map_insert_entry(new_map, new_index, entry, &hash_map->impl->allocator);
+          }
+        }
+      }
+    }
+
+    // Something went wrong above after we allocated the new map. Try to clean it up
+    if (RCUTILS_RET_OK != ret) {
+      hash_map_deallocate_map(new_map, new_capacity, &hash_map->impl->allocator, false);
+      return ret;
+    }
+
+    // Cleanup the old map and swap in the new one
+    ret = hash_map_deallocate_map(hash_map->impl->map, hash_map->impl->capacity,
+        &hash_map->impl->allocator, false);
+    // everything worked up to this point, so if we fail to dealloc the old map still set the new
+    hash_map->impl->map = new_map;
+    hash_map->impl->capacity = new_capacity;
+  }
+
+  return ret;
+}
+
+rcutils_ret_t
+rcutils_hash_map_init(
+  rcutils_hash_map_t * hash_map,
+  size_t initial_capacity,
+  size_t key_size,
+  size_t data_size,
+  rcutils_hash_map_key_hasher_t key_hashing_func,
+  rcutils_hash_map_key_cmp_t key_cmp_func,
+  const rcutils_allocator_t * allocator)
+{
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(key_hashing_func);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(key_cmp_func);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(allocator);
+  if (1 > initial_capacity) {
+    RCUTILS_SET_ERROR_MSG("initial_capacity cannot be less than 1");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  } else if (1 > key_size) {
+    RCUTILS_SET_ERROR_MSG("key_size cannot be less than 1");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  } else if (1 > data_size) {
+    RCUTILS_SET_ERROR_MSG("data_size cannot be less than 1");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
```

Comment:
@nburek same comments in `rcutil_array_list_init()` apply.

Reply:
As I mentioned above, it's a style preference for me. If it violates a style guide then I can change it.